sql: chunk deletions

tbg opened the pull request:
In the absence of a fast path deletion, `DELETE` would generate one
potentially giant batch and OOM the gateway node. This became obvious
quickly via heap profiling.

Added chunking of the deletions to `tableDeleter`. SQL folks may have
stronger opinions on how to achieve this, or a better idea of a
preexisting chunking mechanism that works more reliably. If nothing
else, this change serves as a prototype to fix #17921.

With this change, `roachtest run drop` works (as in, it doesn't
out-of-memory right away; the run takes a long time so I can't yet
confirm that it actually passes).

Release note (sql change): deleting many rows at once now consumes less
memory.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22991)
<!-- Reviewable:end -->

knz commented on pkg/sql/tablewriter.go:
```diff
@@ -736,6 +737,15 @@ func (td *tableDeleter) init(txn *client.Txn, evalCtx *tree.EvalContext) error {
 func (td *tableDeleter) row(
 	ctx context.Context, values tree.Datums, traceKV bool,
 ) (tree.Datums, error) {
+	if td.batchSize > 10000 {
```
extract the constant in the global scope and give it an explanatory comment.

knz reviewed (approved):
LGTM with nit

knz commented:
if this merges into master, please also cherry-pick into 2.0

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/tablewriter.go, line 740 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22991#-L60IBCvGOEhwsQl7MwQ-r1-740:-L6276rcKerj4SbMSkxM:b-roebv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca55a93e81193bf17733b469f13aaca26019a719/pkg/sql/tablewriter.go#L740)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

extract the constant in the global scope and give it an explanatory comment.
</blockquote></details>

Not sure the constant needs to be in the global scope given it is only accessed here, but :thumbsup: on making a constant.

Also, shouldn't this be `>=`? Otherwise you'll be running batches of 10001 rows, which while not wrong, offends my sense of propriety.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22991#-:-L627yRhJLmEcwKZCVpw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
sql: chunk deletions

In the absence of a fast path deletion, `DELETE` would generate one
potentially giant batch and OOM the gateway node. This became obvious
quickly via heap profiling.

Added chunking of the deletions to `tableDeleter`. SQL folks may have
stronger opinions on how to achieve this, or a better idea of a
preexisting chunking mechanism that works more reliably. If nothing
else, this change serves as a prototype to fix #17921.

With this change, `roachtest run drop` works (as in, it doesn't
out-of-memory right away; the run takes a long time so I can't yet
confirm that it actually passes).

Release note (sql change): deleting many rows at once now consumes less
memory.

spencerkimball commented:
@tschottdorf (+@knz, @andreimatei) when I ran this locally, I noticed that we seem to be scanning and buffering the entire set of rows to be deleted. It seems with this change that the total memory footprint should not soar into the GiBs. Why doesn't SQL consume rows from the scanner in a streaming fashion to feed to the table writer?

andreimatei commented:
Spencer is this what you're looking for?
https://github.com/cockroachdb/cockroach/issues/16180

On Feb 24, 2018 1:56 PM, "Spencer Kimball" <notifications@github.com> wrote:

> @tschottdorf <https://github.com/tschottdorf> (+@knz
> <https://github.com/knz>, @andreimatei <https://github.com/andreimatei>)
> when I ran this locally, I noticed that we seem to be scanning and
> buffering the entire set of rows to be deleted. It seems with this change
> that the total memory footprint should not soar into the GiBs. Why doesn't
> SQL consume rows from the scanner in a streaming fashion to feed to the
> table writer?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22991#issuecomment-368250872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcTG21dwJAhCTJST8yPRyKV7CxcECks5tYFtagaJpZM4SQbsB>
> .
>
